### ISSUE TITLE
False positive `Type alias cannot refer to itself indirectly in its definition` when using dataclasses, NamedTuple or NewType

### ISSUE THREAD

**Issue by phaker** [microsoft/pyright#4770]
**Describe the bug**
I'm getting a false positive `Type alias ... cannot refer to itself indirectly in its definition` error on a type alias whose definition refers to a dataclass (or typing.NamedTuple or Newtype).

The exact same line of code triggered #4596 about a month ago, but that one was fixed in 1.1.294 released about a day after I stumbled into it so I just updated and it went away. Now reappeared after minor changes to the types involved.

**Screenshots or Code**

```python
from dataclasses import dataclass
from typing import TypeAlias, NamedTuple, NewType


AnAlias: TypeAlias = "int|Y"


def f(a: AnAlias) -> None:
    pass


f(0)


# # WORKS
# class Y:
#     pass


@dataclass
class Y:
    pass


# @dataclass(init=False, repr=False, eq=False, match_args=False)
# class Y:
#    def __init__(self):
#        pass


# Y = NewType('Y', str)


# Y = NamedTuple('Y', [])


# class Y(NamedTuple):
#     pass
```

Which gives me the error: `Type alias "AnAlias" cannot refer to itself indirectly in its definition (reportGeneralTypeIssues)`.

**VS Code extension or command-line**
Running 1.1.298, with or without `strict`. Pyright finds no other problems with the code.

**Additional context**

* In the the repro below the alias is a union, but in the original code it was a Callable, neither works.

* Replacing the dataclass with an equivalent hand-crafted class makes it work, other commented out examples break.

* The alias needs to be "used" before the class is defined, but there's no need to make a value of the alias type before it's fully defined like the `f(0)` above. This also triggers it:
    ~~~python
    def f() -> Callable[[AnAlias], None]:
        def g(a: AnAlias) -> None:
            pass
        return g


    h = f()
    ~~~

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.299, which I just published. It will also be included in a future release of pylance.

**Comment by phaker** [comment]
Great. Updated and this fixes the problem.

In the rat's nest I found it in too.
